File: medlens/chat/session.py
"""State container for MedLens terminal chat."""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import Literal

from medlens.tools.local_safety import MedicationSafetyReport, NormalizedMedication


@dataclass
class ToolCallRecord:
    name: str
    args: dict[str, object]
    result: dict[str, object] | list[object] | None = None
    error: str | None = None
    duration_ms: int = 0


@dataclass
class ChatSession:
    provider_name: str
    provider_model: str = "template"
    privacy_mode: Literal["on_device", "cloud"] = "on_device"
    medications: list[NormalizedMedication] = field(default_factory=list)
    transcript: list[dict[str, str]] = field(default_factory=list)
    last_report: MedicationSafetyReport | None = None
    last_trace: list[ToolCallRecord] = field(default_factory=list)
    pending_unclear_medications: tuple[str, ...] = ()
    pending_clear_medications: tuple[str, ...] = ()

    def medication_inputs(self) -> tuple[str, ...]:
        return tuple(item.input_name for item in self.medications)

    def resolved_canonicals(self) -> tuple[str, ...]:
        seen: set[str] = set()
        values: list[str] = []
        for item in self.medications:
            if not item.resolved or item.canonical_name is None:
                continue
            if item.canonical_name in seen:
                continue
            seen.add(item.canonical_name)
            values.append(item.canonical_name)
        return tuple(values)

    def clear_turn_trace(self) -> None:
        self.last_trace = []

    def clear_pending_clarification(self) -> None:
        self.pending_unclear_medications = ()
        self.pending_clear_medications = ()

File: web/src/chat/session.ts
// Chat session state — TS port of `medlens/chat/session.py`. Holds the in-memory
// medication list, transcript, last deterministic report, and tool-call trace
// for the current turn. Persistence (sidebar / multi-conversation) is handled
// separately in Phase 6 via Zustand + idb-keyval.

import type { MedicationSafetyReport, NormalizedMedication } from "../tools/types";

export interface ToolCallRecord {
  name: string;
  args: Record<string, unknown>;
  result?: unknown;
  error?: string;
  duration_ms?: number;
}

export interface TranscriptMessage {
  role: "user" | "assistant" | "tool";
  content: unknown;
  tool_calls?: { id: string; name: string; args: Record<string, unknown> }[];
  tool_call_id?: string;
  name?: string;
}

export class ChatSession {
  medications: NormalizedMedication[] = [];
  transcript: TranscriptMessage[] = [];
  last_report: MedicationSafetyReport | null = null;
  last_trace: ToolCallRecord[] = [];
  provider_name = "template";
  provider_model: string | null = null;
  privacy_mode: "offline" | "cloud" = "offline";

  constructor(init?: Partial<Pick<ChatSession, "provider_name" | "provider_model" | "privacy_mode">>) {
    if (init?.provider_name) this.provider_name = init.provider_name;
    if (init?.provider_model !== undefined) this.provider_model = init.provider_model;
    if (init?.privacy_mode) this.privacy_mode = init.privacy_mode;
  }

  /** Inputs in stable order for tools that default to "current session meds". */
  medicationInputs(): string[] {
    return this.medications.map((m) => m.input_name);
  }

  clearTurnTrace(): void {
    this.last_trace = [];
  }
}
